contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol

```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
+    // Build minimal valid request data
+    string memory sourceCode = "return 'hello world';";
+    FunctionsRequest.Request memory request;
+    FunctionsRequest.initializeRequest(
+      request,
+      FunctionsRequest.Location.Inline,
+      FunctionsRequest.CodeLanguage.JavaScript,
+      sourceCode
+    );
+    bytes memory requestData = FunctionsRequest.encodeCBOR(request);
+
+    bytes32 incorrectDonId = s_functionsRouter.getAllowListId();
+
+    // Low level revert from incorrect call
+    vm.expectRevert();
+    s_functionsRouter.sendRequest(
+      s_subscriptionId,
+      requestData,
+      FunctionsRequest.REQUEST_DATA_VERSION,
+      5_000,
+      incorrectDonId
+    );
+  }
+
+  function test_SendRequest_RevertIfPaused() public {
+    s_functionsRouter.pause();
+
+    // Build minimal valid request data
+    string memory sourceCode = "return 'hello world';";
+    FunctionsRequest.Request memory request;
+    FunctionsRequest.initializeRequest(
+      request,
+      FunctionsRequest.Location.Inline,
+      FunctionsRequest.CodeLanguage.JavaScript,
+      sourceCode
+    );
+    bytes memory requestData = FunctionsRequest.encodeCBOR(request);
+
+    vm.expectRevert("Pausable: paused");
+    s_functionsRouter.sendRequest(s_subscriptionId, requestData, FunctionsRequest.REQUEST_DATA_VERSION, 5000, s_donId);
+  }
+
+  function test_SendRequest_RevertIfNoSubscription() public {
+    // Build minimal valid request data
+    string memory sourceCode = "return 'hello world';";
+    FunctionsRequest.Request memory request;
+    FunctionsRequest.initializeRequest(
+      request,
+      FunctionsRequest.Location.Inline,
+      FunctionsRequest.CodeLanguage.JavaScript,
+      sourceCode
+    );
+    bytes memory requestData = FunctionsRequest.encodeCBOR(request);
+
+    uint64 invalidSubscriptionId = 123456789;
+
+    vm.expectRevert(FunctionsSubscriptions.InvalidSubscription.selector);
+    s_functionsRouter.sendRequest(
+      invalidSubscriptionId,
+      requestData,
+      FunctionsRequest.REQUEST_DATA_VERSION,
+      5000,
+      s_donId
+    );
+  }
+
+  function test_SendRequest_RevertIfConsumerNotAllowed() public {
+    // Remove sending wallet as a subscription consumer
+    s_functionsRouter.removeConsumer(s_subscriptionId, OWNER_ADDRESS);
 
+    // Build minimal valid request data
+    string memory sourceCode = "return 'hello world';";
+    FunctionsRequest.Request memory request;
+    FunctionsRequest.initializeRequest(
+      request,
+      FunctionsRequest.Location.Inline,
+      FunctionsRequest.CodeLanguage.JavaScript,
+      sourceCode
+    );
+    bytes memory requestData = FunctionsRequest.encodeCBOR(request);
+
+    vm.expectRevert(FunctionsSubscriptions.InvalidConsumer.selector);
+    s_functionsRouter.sendRequest(s_subscriptionId, requestData, FunctionsRequest.REQUEST_DATA_VERSION, 5000, s_donId);
+  }
+
+  function test_SendRequest_RevertIfInvalidCallbackGasLimit() public {
```

Comment:
Can we add a test case for reversions if the subscription balance isn't sufficient?
We could test both the cases where the user sets the callback gas limit to something reasonable like 100k and the case where the gas limit is zero just to confirm expected behavior.

Reply:
Just added one for gas limit = 0, but for the other one could you explain more what you mean?
Is that not `test_Fulfill_RequestNotProcessedSubscriptionBalanceInvariant`?

Reply:
"do you mean the same test for insufficient sub balance but with two different callbackGasLimits in the request?"

Yes.  I was just thinking of the potential edge case where the gas limit specified is 0, but the user still doesn't have a large enough subscription balance to cover the baseline cost.
If it makes more sense, we can add the tests in FunctionsCoordinator.t.sol.